DD-711: Predict dataverse files from easy_files in frsdb (fedora)

jo-pol opened the pull request:
Fixes DD-711: Predict dataverse files from easy_files in frsdb (fedora)

# Description of changes
 * core functionality in `EasyFileLoader`
 * see also javadoc of  getters at the end of `FedoraToBagCsv` and `EasyFileLoaderTest`

# How to test

* in the root of easy-dtap: `start-preprovisioned-box.py`
* in the root of dd-verify-file-migration: 
  * once: run `start-env.sh` and change the password for the easyDb in `etc/config.yml`
  * per test (each command line in another terminal, or the second command in Intellij by running DdVerifyFileMigrationApplication configured with the specified arguments):

        rm -rf data;start-hsqldb.sh
        start.sh load-from-fedora etc/config.yml src/test/resources/deasy.csv does-not-exist.csv

Or build, deploy on deasy, copy `src/test/resources/deasy.csv` to `easy-dtap/shared` and:

    dd-verify-file-migration load-from-fedora /etc/opt/dans.knaw.nl/dd-verify-file-migration/config.yml /vagrant/shared/deasy.csv

# Related PRs

* deploy with https://github.com/DANS-KNAW/easy-dtap/pull/528

# Notify
@DANS-KNAW/dataversedans

jo-pol committed:
set up DAO's and table mapping

jo-pol committed:
license headers

mderuijter commented on src/main/java/nl/knaw/dans/filemigration/api/EasyFile.java:
```diff
@@ -0,0 +1,125 @@
+@Entity
+@Table(name = "easy_files")
+@NamedQueries({ @NamedQuery(name = "EasyFile.findByDatasetId",
+                            query = "SELECT * FROM EasyFile ef WHERE ef.dataset_sid = :datasedSid ORDER BY path"),
```
@jo-pol Here are some examples of NamedQueries and using parameters:
[LicenseServiceBean(or LicenseDAO in Dropwizard)](https://github.com/DANS-KNAW/dataverse/blob/a79143e6b61047e1b61e28b671f03b90bfdc405e/src/main/java/edu/harvard/iq/dataverse/LicenseServiceBean.java#L55-L65)

[License(Entity class)](https://github.com/DANS-KNAW/dataverse/blob/a79143e6b61047e1b61e28b671f03b90bfdc405e/src/main/java/edu/harvard/iq/dataverse/License.java#L21-L39)

jo-pol committed:
review @mderuijter

jo-pol committed:
logging and debug-etc

jo-pol commented on src/test/resources/debug-etc/config.yml:
```diff
@@ -27,3 +27,22 @@ logging:
+    logValidationErrors: true
+    properties:
+      hibernate.dialect: 'org.hibernate.dialect.PostgreSQL92Dialect'
+      hibernate.hbm2ddl.auto: update
```
This database is used just read only. Do we need this line?

jo-pol replied:
I suppose we also need to configure a user and password.

jo-pol committed:
fix configuration issues

jo-pol commented on src/main/java/nl/knaw/dans/filemigration/db/EasyFileDAO.java:
```diff
@@ -0,0 +1,33 @@
+  }
+
+  public List<EasyFile> findByDatasetId(String id) {
+    return list(namedTypedQuery(EasyFile.FIND_BY_DATASET_ID).setParameter(EasyFile.DATASET_ID, id));
```
Currently throws

    org.hibernate.HibernateException: No session currently bound to execution context

janvanmansum replied:
Have a look at this code: https://github.com/DANS-KNAW/dd-manage-prestaging/blob/master/src/main/java/nl/knaw/dans/prestaging/cli/LoadFromDataverseCommand.java#L61

And also the linked Stackoverflow post. Summary: you need proxy that lets you use the `@UnitOfWork` annotation.

janvanmansum commented on src/main/assembly/dist/cfg/config.yml:
```diff
@@ -11,25 +11,22 @@ server:
+  properties:
+    hibernate.dialect: 'org.hibernate.dialect.PostgreSQL92Dialect'
+    hibernate.hbm2ddl.auto: update
+    user: fedora_db_admin
```
`user` and `password` should be directly under `easyDb`, not under `properties`.

janvanmansum commented on src/main/assembly/dist/cfg/config.yml:
```diff
@@ -11,25 +11,22 @@ server:
+    password: changeme
 
-verificationDatabase:
+fileMigrationDb:
```
Not sure why you consider this a better name.

janvanmansum commented on src/main/java/nl/knaw/dans/filemigration/api/Expected.java:
```diff
@@ -0,0 +1,117 @@
+  private String sha1checksum;
+  @Column(name="easy_file_id")
+  private String easyFileId;
+  @Column(name="fsRdbPath")
```
Why not keep up the snake_case convention for column names; you are switching to camelCase for some reason...?

jo-pol committed:
fix: No session currently bound to execution context

jo-pol committed:
config details

jo-pol committed:
snake_case as reviewed, hardcoded query

jo-pol committed:
down to native query

jo-pol commented on src/main/java/nl/knaw/dans/filemigration/core/EasyFileLoader.java:
```diff
@@ -0,0 +1,49 @@
+  @UnitOfWork
+  public void loadFromDatasetId(String id) {
+    for (Object ef : dao.findByDatasetId("easy-dataset:9")) {
+      log.info("ef = {}" , ef);
```
I'm clearly blind for something. `trace` does not work despite 
https://github.com/DANS-KNAW/dd-verify-file-migration/blob/752502403c8fd0407cd8ebe0875718791a5ab176/src/test/resources/debug-etc/config.yml#L17-L22

also tried the following in `etc/config.yml` (which otherwise equals the file above except for the DB password) what did produce the executed query but `log.isTracelLevel()` stubbornly returns false here.
```
  loggers:
    "io.dropwizard": INFO
    "org.hibernate.SQL": DEBUG
    "dans.knaw.nl": TRACE
```

jo-pol replied:
Solved in https://github.com/DANS-KNAW/dans-module-archetype/commit/13335d609f9f20a0b627cbd396a0a3280f78a4e4

jo-pol committed:
fix trace logging

jo-pol committed:
parameterize the query

jo-pol commented on src/main/java/nl/knaw/dans/filemigration/db/EasyFileDAO.java:
```diff
@@ -0,0 +1,41 @@
+  public List<EasyFile> findByDatasetId(String id) {
+    // TODO createNamedQuery(EasyFile.FIND_BY_DATASET_ID, EasyFile.class)
+    Session session = currentSession();
+    Query<EasyFile> query = session.createSQLQuery("SELECT * FROM easy_files WHERE dataset_sid = ? ORDER BY path");
```
Changing this line into 

    Query<EasyFile> query = session.createQuery("SELECT f FROM nl.knaw.dans.filemigration.api.EasyFile f WHERE dataset_sid = ? ORDER BY f.path", EasyFile.class).setReadOnly( true );

causes

```
java.lang.ArrayIndexOutOfBoundsException: 0
	at org.hibernate.internal.AbstractSharedSessionContract.resultClassChecking(AbstractSharedSessionContract.java:833)
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:817)
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:113)
	at nl.knaw.dans.filemigration.db.EasyFileDAO.findByDatasetId(EasyFileDAO.java:35)
```
apparently there is no queryPlan.getTranslators()

	protected void resultClassChecking(Class resultClass, org.hibernate.Query hqlQuery) {
		// make sure the query is a select -> HHH-7192
		final HQLQueryPlan queryPlan = getFactory().getQueryPlanCache().getHQLQueryPlan(
				hqlQuery.getQueryString(),
				false,
				getLoadQueryInfluencers().getEnabledFilters()
		);
		if ( queryPlan.getTranslators()[0].isManipulationStatement() )

mderuijter replied:
@jo-pol The HSQL Query seems off to me:
```
"SELECT f FROM nl.knaw.dans.filemigration.api.EasyFile f WHERE dataset_sid = ? ORDER BY f.path"
```
Should be:
```
"SELECT f FROM nl.knaw.dans.filemigration.api.EasyFile f WHERE f.datasetSid = :datasetSid ORDER BY f.datasetSid"
```

mderuijter commented on src/main/java/nl/knaw/dans/filemigration/DdVerifyFileMigrationApplication.java:
```diff
@@ -19,8 +19,20 @@
+import nl.knaw.dans.filemigration.db.EasyFileDAO;
 
 public class DdVerifyFileMigrationApplication extends Application<DdVerifyFileMigrationConfiguration> {
+    private final HibernateBundle<DdVerifyFileMigrationConfiguration> hibernate = new HibernateBundle<DdVerifyFileMigrationConfiguration>(Expected.class) {
```
@jo-pol add all entities separated by a comma:
```
new HibernateBundle<DdVerifyFileMigrationConfiguration>(Expected.class)
```
Should be:
```
new HibernateBundle<DdVerifyFileMigrationConfiguration>(Expected.class, EasyFile.class)
```

jo-pol committed:
fixed hybernate bundle

jo-pol committed:
fixed hybernate bundle (the rest)

jo-pol committed:
snake_case for Expected

jo-pol committed:
read IDs from CSVs

jo-pol committed:
CSV class, skip FAILEd records

jo-pol committed:
add migration files

jo-pol committed:
license header

jo-pol commented on src/main/java/nl/knaw/dans/filemigration/core/FedoraToBagCsv.java:
```diff
@@ -0,0 +1,92 @@
+  @Override
+  public String toString() {
+    // TODO generated by IntelliJ but do we want this?
+    return "FedoraToBagCsv{" + "datasetId='" + datasetId + '\'' + ", doi='" + doi + '\'' + ", comment='" + comment + '\'' + ", type='" + type + '\'' + ", r=" + r + '}';
```
What are our best practices for toString methods? See also 
https://github.com/DANS-KNAW/dd-verify-file-migration/blob/f5568db777caf5bec821a8cb919dd7ba8efb30d5/src/main/java/nl/knaw/dans/filemigration/api/EasyFile.java#L68
https://github.com/DANS-KNAW/dd-verify-file-migration/blob/c845951759c3f0f6840fc345d504091a434b6f89/src/main/java/nl/knaw/dans/filemigration/api/Expected.java#L48

jo-pol committed:
prepare for another hybernate bundle

jo-pol committed:
as discussed with @ lindareijnhoudt

jo-pol committed:
no migration-info-db but retrieve files with dataverse-api

jo-pol committed:
dotting i-s: order, unique, retries

jo-pol committed:
attepmt to fix: org.hibernate.MappingException: Unknown entity: nl.knaw.dans.filemigration.api.Expected

jo-pol committed:
attepmt to fix bundles

jo-pol committed:
use bundle name in UnitOfWork

jo-pol committed:
configure DB with alias

jo-pol committed:
exception handling at proper level

jo-pol committed:
renamed too many, other cleanup

jo-pol committed:
rename Expected -> ExpectedFile

jo-pol committed:
using IdClass

jo-pol committed:
license header

jo-pol committed:
first unit tests

jo-pol committed:
fixed some not yet used mocking

jo-pol committed:
mockable EasyFileLoader

jo-pol committed:
ignore files created by start-hsqldb.sh

jo-pol committed:
fix test with equals/hash method

jo-pol committed:
unit test for duplicates

jo-pol committed:
decreasing order of easy-file:NNN

jo-pol committed:
duplicate constraint (primary key) as specified

jo-pol committed:
table creation script

jo-pol committed:
fix DB deployment

jo-pol commented on src/main/assembly/dist/cfg/config.yml:
```diff
@@ -11,25 +11,22 @@ server:
-    properties:
-      hibernate.dialect: 'org.hibernate.dialect.HSQLDialect'
-      hibernate.hbm2ddl.auto: update
+  url: 'jdbc:postgresql://easy.dans.knaw.nl:5432/dd_verify_file_migration'
```
Somewhere must be a mistake, after deploy and running on deasy (as specified in the cover message of this PR) I get a connection error, the stack trace does not give a clue on which DB.

     org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator: HHH000342: Could not obtain connection to query metadata : The connection attempt failed.

jo-pol replied:
localhost in the URL does improve things.

jo-pol committed:
fix DB deployment

jo-pol commented on src/main/assembly/dist/install/db-tables.sql:
```diff
@@ -0,0 +1,14 @@
+ transformed_name boolean,
+ primary key (doi, expected_path, removed_duplicate_file_count)
+);
+GRANT INSERT, SELECT, UPDATE, DELETE ON expected TO dd_verify_file_migration;
```
I did not see this working when deploying, but manual creation with the same content works.

jo-pol commented on src/main/java/nl/knaw/dans/filemigration/core/EasyFileLoader.java:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  public void loadFromCsv(FedoraToBagCsv csv) {
+    if (csv.getComment().contains("OK") && !csv.getComment().contains("no payload"))
```
The nopayload datasets do create migration files.

jo-pol committed:
deploy fix

jo-pol committed:
expect migration file for datasets without payload

jo-pol committed:
mandatory arguments

usage change from
    load-from-fedora [-h] [file] [csv [csv ...]]
to
    load-from-fedora [-h] file csv [csv ...]

jo-pol commented on src/main/java/nl/knaw/dans/filemigration/cli/LoadFromFedoraCommand.java:
```diff
@@ -0,0 +1,92 @@
+    public void configure(Subparser subparser) {
+        // mandatory variant of: super.configure(subparser);
+        subparser.addArgument("file")
+            .type(File.class)
+            .required(true)
```
An alternative could be

    .setDfeault("/etc/opt/dans.knaw.nl/dd-verify-file-migration/config.yml");

It feels a bit hacky, relying on https://github.com/dropwizard/dropwizard/blob/dd64c768b6058c18484bce18753359846d5cffae/dropwizard-core/src/main/java/io/dropwizard/cli/ConfiguredCommand.java#L88

I would rather have additional arguments for `super.configure` to be more flexible. I was also lucky to choose "csv" (and not "file") as argument.

jo-pol committed:
redundant call

jo-pol commented on src/main/java/nl/knaw/dans/filemigration/core/EasyFileLoader.java:
```diff
@@ -0,0 +1,139 @@
+        saveExpected(expected);
+      } catch(PersistenceException e){
+        // logged as error by org.hibernate.engine.jdbc.spi.SqlExceptionHelper
+        if (!(e.getCause() instanceof ConstraintViolationException))
```
In theory this can also be a "NOT NULL" violation, not just a duplicate key. The retry is not wrapped in a try catch, not sure we need a third attempt, if so, that will fail.

lindareijnhoudt commented on src/main/java/nl/knaw/dans/filemigration/core/FedoraToBagCsv.java:
```diff
@@ -0,0 +1,91 @@
+    datasetId = r.get(DATASET_ID_COLUMN);
+    doi = r.get(DOI_COLUMN);
+    comment = r.get(COMMENT_COLUMN);
+    type = r.get(TRANSFORMATION_TYPE_COLUMN);
```
```suggestion
    transformation = r.get(TRANSFORMATION_TYPE_COLUMN);
```
`transformation` is a more meaningful name than `type`

lindareijnhoudt commented on src/main/java/nl/knaw/dans/filemigration/core/EasyFileLoader.java:
```diff
@@ -0,0 +1,139 @@
+    expected.setFs_rdb_path(ef.getPath());
+    expected.setExpected_path(dvPath);
+    expected.setAdded_during_migration(false);
+    expected.setRemoved_thumbnail(path.matches(".*thumbnails/.*_small.(png|jpg|tiff)"));
```
should we account for uppercase?

lindareijnhoudt replied:
i thought to have expected_path only filled for files that we expect in the dataverse dataset. I guess that is not possible when the expected_path is part of the primary key. so this implementation means we need to filter on `removed_thumbnail = false`

lindareijnhoudt commented on src/main/java/nl/knaw/dans/filemigration/core/EasyFileLoader.java:
```diff
@@ -0,0 +1,139 @@
+    }
+  }
+
+  /** note: bag-to-deposit also adds emd.xml to bags from the vault, that does not apply to migration */
```
```suggestion
  /** note: convert-bag-to-deposit also adds amd.xml to bags from the vault, in the metadata-folder, that does not apply to the payload easy-migration files
  For Fedora datasets, the migration-files also include the emd.xml */
```

lindareijnhoudt commented on src/main/java/nl/knaw/dans/filemigration/core/EasyFileLoader.java:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  /** note: bag-to-deposit also adds emd.xml to bags from the vault, that does not apply to migration */
+  private static final String[] migrationFiles = { "provenance.xml", "dataset.xml", "files.xml" };
```
```suggestion
  private static final String[] migrationFiles = { "provenance.xml", "dataset.xml", "files.xml", "emd.xml" };
```

jo-pol replied:
Whoops:
https://github.com/DANS-KNAW/easy-convert-bag-to-deposit/blob/ec42d59bd7e85be0ebe48f0a53d4c65e8369436b/src/main/scala/nl.knaw.dans.easy.bag2deposit/EasyConvertBagToDepositApp.scala#L186

jo-pol replied:
anyway, we are handling files from fedora in this class, not from the vault.

lindareijnhoudt commented on src/main/java/nl/knaw/dans/filemigration/core/EasyFileLoader.java:
```diff
@@ -0,0 +1,139 @@
+    expected.setSha1_checksum("");
+    expected.setEasy_file_id("");
+    expected.setFs_rdb_path("");
+    expected.setExpected_path("migration/" + migrationFile);
```
```suggestion
    expected.setExpected_path("easy-migration/" + migrationFile);
```

jo-pol committed:
as reviewed

https://github.com/DANS-KNAW/dd-verify-file-migration/pull/2#pullrequestreview-827493574

jo-pol committed:
fix broken test
